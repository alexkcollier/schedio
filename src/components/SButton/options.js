export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
