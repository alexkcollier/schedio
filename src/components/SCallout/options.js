export const calloutTypes = ['', 'success', 'warning', 'error']
