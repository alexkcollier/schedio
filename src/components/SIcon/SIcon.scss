.icon {
  fill: none;
  height: 1.5em;
  stroke: currentColor;
  stroke-linecap: round;
  stroke-linejoin: round;
  // using px because android browsers render stroke width twice as thick when using ems
  stroke-width: 2px;
  width: 1.5em;

  &--size-small {
    height: 0.75em;
    width: 0.75em;
  }

  &--size-regular {
    height: 1.5em;
    width: 1.5em;
  }

  &--size-large {
    height: 4em;
    width: 4em;
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      stroke: color($color-name);
    }
  }
}
