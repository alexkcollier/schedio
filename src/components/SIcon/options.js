export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options';

export const sizes = [
  'small',
  'regular',
  'large',
];
