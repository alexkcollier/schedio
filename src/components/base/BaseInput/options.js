export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
