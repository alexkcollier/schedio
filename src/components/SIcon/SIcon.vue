<template>
  <!-- eslint-disable vue/no-v-html -->
  <svg
    :class="classList"
    class="icon"
    viewBox="0 0 24 24"
    v-html="featherIcon.contents"
  />
</template>

<script>
import { colors, sizes } from './options.js'
import feather from 'feather-icons'
import icons from 'feather-icons/dist/icons.json'

const list = Object.keys(icons)

export default {
  name: 'SIcon',

  props: {
    icon: {
      type: String,
      required: true,
      validator: v => {
        return list.includes(v) || console.error(`\`icon\` must be one of: ${list.join(', ')}`)
      }
    },

    iconColor: {
      type: String,
      default: '',
      validator: value => {
        if (!value || colors.includes(value)) return true

        return console.error(`\`iconColor\` ${value} not found. Allowed colors: ${colors}`)
      }
    },

    iconSize: {
      type: String,
      default: '',
      validator: value => {
        if (!value || sizes.includes(value)) return true

        return console.error(`\`iconSizes\` ${value} not found. Allowed sizes: ${sizes}`)
      }
    }
  },

  computed: {
    classList() {
      return [
        this.iconColor ? `icon--color-${this.iconColor}` : '',
        this.iconSize ? `icon--size-${this.iconSize}` : ''
      ]
    },

    featherIcon() {
      return feather.icons[this.icon] || { contents: '' }
    }
  }
}
</script>
