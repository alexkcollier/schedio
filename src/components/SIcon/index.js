import SIcon from './SIcon.vue';

export { SIcon };

export default {
  install: Vue => Vue.component(SIcon.name, SIcon),
};
