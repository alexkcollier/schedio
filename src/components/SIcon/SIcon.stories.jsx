import { colors, sizes } from '@/components/SIcon/options'
import { SHeading } from '@/components/SHeading'
import { SIcon } from '@/components/SIcon'
import PropList from '@@/docs/components/PropList'
import { select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import icons from 'feather-icons/dist/icons.json'

const colorOptions = ['', ...colors]
const sizeOptions = ['', ...sizes]
const iconList = Object.keys(icons)

storiesOf('Components/SIcon', module)
  .addParameters({ jest: 'SIcon' })
  .add('Icon', () => {
    return {
      props: {
        props: {
          default: () => ({
            icon: select('icon', iconList, iconList[0], 'Required Props'),
            color: select('color', colorOptions, '', 'Optional Props'),
            size: select('size', sizeOptions, '', 'Optional Props'),
          }),
        },
      },
      render (h) {
        const { props } = this.$props
        return (
          <div>
            <SHeading level="1">Icon</SHeading>
            <p>
              The icon component uses{' '}
              <a href="https://feathericons.com/" class="link link--external">
                feather-icons
              </a>{' '}
              to provide icons.
            </p>

            <SHeading level="2">Example</SHeading>
            <div style="align-items: center; display: flex; padding: 0.5em 0;">
              <SIcon {...{ props }} />
              <div style="padding-left: 0.5em;">
                <div>
                  Icon name: <code>{props.icon}</code>
                </div>
                {props.color && (
                  <div>
                    Icon color: <code>{props.color}</code>
                  </div>
                )}
                {props.size && (
                  <div>
                    Icon size: <code>{props.size}</code>
                  </div>
                )}
              </div>
            </div>

            {SIcon.props && <PropList component={SIcon} />}
          </div>
        )
      },
    }
  })
  .add(
    'All icons',
    () => ({
      render (h) {
        return (
          <div>
            <SHeading>All icons</SHeading>
            <p>Icons are available in all of Sch√©dio's base colors</p>

            <div style="display: flex; flex-wrap: wrap;">
              {iconList.map((icon) => {
                return (
                  <div style="align-items: center; display: flex; flex-direction: column; padding: 1em; width: 25%">
                    <SIcon icon={icon} />
                    {icon}
                  </div>
                )
              })}
            </div>
          </div>
        )
      },
    }),
    {
      options: {
        showPanel: false,
      },
    }
  )
