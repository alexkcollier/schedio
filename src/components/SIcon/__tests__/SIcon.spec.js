import { shallowMount } from '@vue/test-utils'
import SIcon from '@/components/SIcon/SIcon.vue'
import * as options from '@/components/SIcon/options'
import icons from 'feather-icons/dist/icons.json'

const iconList = Object.keys(icons)

describe('SIcon.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { })
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SIcon, {
      propsData: {
        icon: iconList[0],
      },
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(errorSpy).not.toHaveBeenCalled()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('requires an icon name', () => {
    shallowMount(SIcon)
    expect(errorSpy).toHaveBeenCalled()
  })

  describe.each(Object.keys(options.colors))('color "%s"', (color) => {
    it('can be the base color', async () => {
      wrapper.setProps({ color })
      await wrapper.vm.$nextTick()
      expect(wrapper.find(`.icon--color-${color}`).exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    if (!options.colors[color].length) return

    it.each(options.colors[color])('can be the shade "%s"', async (shade) => {
      wrapper.setProps({ color, shade })
      await wrapper.vm.$nextTick()
      expect(wrapper.find(`.icon--color-${color}-${shade}`).exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('validates', () => {
    it('icons', () => {
      shallowMount(SIcon, { propsData: { icon: 'not an icon' } })
      expect(errorSpy).toHaveBeenCalled()
    })

    it('color', () => {
      shallowMount(SIcon, { propsData: { icon: iconList[0], color: 'not a color' } })
      expect(errorSpy).toHaveBeenCalled()
    })

    it('shade', () => {
      shallowMount(SIcon, {
        propsData: {
          icon: iconList[0],
          color: Object.keys(options.colors)[0],
          shade: 'not a shade',
        },
      })
      expect(errorSpy).toHaveBeenCalledTimes(2)
    })

    it('size', () => {
      shallowMount(SIcon, { propsData: { icon: iconList[0], size: 'not a size' } })
      expect(errorSpy).toHaveBeenCalled()
    })
  })

  describe('size', () => {
    it.each(options.sizes)('can be "%s"', async (size) => {
      wrapper.setProps({ size })
      await wrapper.vm.$nextTick()
      expect(wrapper.find(`.icon--size-${size}`).exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
