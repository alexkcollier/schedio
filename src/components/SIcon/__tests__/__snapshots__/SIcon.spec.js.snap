// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SIcon.vue color "blue" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-blue">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "gold" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-gold">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "green" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-green">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "grey" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-grey">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "ice" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-ice">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "magenta" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-magenta">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "night" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-night">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "red" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-red">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "turquoise" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-turquoise">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue color "white" can be the base color 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--color-white">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue renders correctly 1`] = `
<svg viewBox="0 0 24 24" class="icon">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue size can be "large" 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--size-large">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue size can be "regular" 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--size-regular">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;

exports[`SIcon.vue size can be "small" 1`] = `
<svg viewBox="0 0 24 24" class="icon icon--size-small">
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12"></polyline>
</svg>
`;
